packages/edit-widgets/src/components/customizer-edit-widgets-initializer/sync-customizer.js

```diff
@@ -0,0 +1,120 @@
+
+- On edit, the client updates the 2-way bound input with a new object that maps
+widget area IDs and the values are serialized block content, encoded
+as a JSON string.
+
+- On publish, a PHP action will parse the JSON string in the
+changeset and update all the widget areas in it, to store the
+new content.
+*/
+
+// Get widget areas from the store in an `id => blocks` mapping.
+const getWidgetAreasObject = () => {
+	const { getWidgetAreas, getBlocksFromWidgetArea } = window.wp.data.select(
+		'core/edit-widgets'
+	);
+
+	return getWidgetAreas().reduce( ( widgetAreasObject, { id } ) => {
+		widgetAreasObject[ id ] = getBlocksFromWidgetArea( id );
+		return widgetAreasObject;
+	}, {} );
+};
+
+// Serialize the provided blocks and render them in the widget area with the provided ID.
+const previewBlocksInWidgetArea = throttle( ( id, blocks ) => {
+	const customizePreviewIframe = document.querySelector(
+		'#customize-preview > iframe'
+	);
+	if ( ! customizePreviewIframe || ! customizePreviewIframe.contentDocument ) {
+		return;
+	}
+
+	const widgetArea = customizePreviewIframe.contentDocument.querySelector(
+		`[data-customize-partial-placement-context*='"sidebar_id":"${ id }"']`
+	);
+	if ( widgetArea ) {
+		widgetArea.innerHTML = serialize( blocks );
+	}
+}, 1000 );
+
+// Update the hidden input that has 2-way data binding with Customizer settings.
+const updateSettingInputValue = throttle( ( nextWidgetAreas ) => {
+	const settingInput = document.getElementById(
+		'_customize-input-gutenberg_widget_blocks'
+	);
+	if ( settingInput ) {
+		settingInput.value = JSON.stringify(
+			Object.keys( nextWidgetAreas ).reduce( ( value, id ) => {
+				value[ id ] = serialize( nextWidgetAreas[ id ] );
+				return value;
+			}, {} )
+		);
+		settingInput.dispatchEvent( new window.Event( 'change' ) );
+	}
+}, 1000 );
+
+// Check that all the necessary globals are present.
+if ( window.wp && window.wp.customize && window.wp.data ) {
+	// Wait for the Customizer to finish bootstrapping.
+	window.wp.customize.bind( 'ready', () =>
+		window.wp.customize.previewer.bind( 'ready', () => {
+			// Try to parse a previous changeset from the hidden input.
+			let widgetAreas;
+			try {
+				widgetAreas = JSON.parse(
+					document.getElementById( '_customize-input-gutenberg_widget_blocks' )
+						.value
+				);
+				widgetAreas = Object.keys( widgetAreas ).reduce( ( value, id ) => {
+					value[ id ] = parse( widgetAreas[ id ] );
+					return value;
+				}, {} );
+			} catch ( err ) {}
+			// This will setup widget areas as normal, but content for a sidebar
+			// that we pass here will take precedence over its published content.
+			window.wp.data.dispatch( 'core/edit-widgets' ).setupWidgetAreas( widgetAreas );
+
+			// Subscribe to registry changes so that we can update the preview
+			// and the hidden input's value when any of the widget areas change.
+			widgetAreas = getWidgetAreasObject();
+			window.wp.data.subscribe( () => {
```

Comment:
Are we considered in terms of performance here? I feel we don't really need to support the live-preview, especially because the rendered HTML can be a lot different from the HTML used in the blocks. We can have dynamic blocks, we can have server-side hooks... I feel it's just going to be confusing. 

That said, I don't think it's a blocker for this PR.

Reply:
>Are we considered in terms of performance here?...

I think that with the throttling in place, the performance hit is minimal, although I haven't done any serious testing besides 6x CPU slowdown on my machine working fine.

>...especially because the rendered HTML can be a lot different from the HTML used in the blocks...

I have plans for another PR that adds an asynchronous `serialize` that uses `<ServerSideRender />` to render dynamic blocks. I could also look at time-slicing or offloading serialization to another thread and that would also take care of any performance concerns here.